File: leetCode/08rainWater.c

int trap(int* height, int heightSize){
    
    int result = 0;

    int currentMaximum = 0;
    int tmp = 0;
    
    int i = 0;
    int bar = 0;

    for (i=0; i<heightSize; i++) {
        bar = height[i];
            
        if (bar >= currentMaximum) {
            result +=  tmp;

            currentMaximum = bar ;
            tmp = 0;
        } else {
            tmp += currentMaximum - bar;
        }
    }
    
    currentMaximum = 0;
    tmp = 0;
    
    for (i=heightSize - 1; i>=0; i--) {
        bar = height[i];

        if (bar > currentMaximum) {
            result +=  tmp;

            currentMaximum = bar ;
            tmp = 0;
        } else {
            tmp += currentMaximum - bar;
        }
    }
    return result;
}

File: leetCode/08rainWater.py
class Solution:
    def trap(self, height: list[int]) -> int:

        result = 0

        currentMaximum = 0
        tmp = 0

        for leftBar in height:
            if leftBar >= currentMaximum:
                result +=  tmp

                currentMaximum = leftBar
                tmp = 0
            else:
                tmp += currentMaximum - leftBar

        currentMaximum = 0
        tmp = 0
        for rightBar in height[::-1]:
            if rightBar > currentMaximum:
                result +=  tmp

                currentMaximum = rightBar 
                tmp = 0
            else:
                tmp += currentMaximum - rightBar



        return result


sol = Solution()
h=[0,1,0,2,1,0,1,3,2,1,2,1] # 6
print(sol.trap(h))
h=[4,2,0,3,2,5] # 9
print(sol.trap(h))
